Implement `dup_series_*` Functions for Series Manipulation

Jatinbhardwaj-093 opened the pull request:
#### Brief description of what is fixed or changed
All series functions perform operations on the input polynomials and return the result modulo `x**prec`. These functions support development of the new poly/series module.

This is needed for gh-28109

`dup_truncate` returns the first n terms (lowest-degree terms) of a polynomial, unlike `dup_slice`, which returns a general slice.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* polys
  * Added dup_series_reversion, dup_series_compose in `densetools.py`.
  * Added dup_series_mul, dup_series_pow in `densearith.py`.
  * Added dup_truncate, dup_from_list, dup_print in `densebasic.py`.
<!-- END RELEASE NOTES -->

Jatinbhardwaj-093 committed:
added dup_reversion to comput compositional inverse

Jatinbhardwaj-093 committed:
fix docs

Jatinbhardwaj-093 commented:
> Is this expected:
In [34]: from sympy.polys.densetools import dup_reversion, dup_compose
In [35]: p1 = [QQ(1), QQ(1), QQ(0)]
In [36]: dup_compose(dup_reversion(p1, 3, QQ), p1, QQ)
Out[36]: [1, 1, 0]

No, I mistakenly truncated the value to `k-1` when it should have been `k`.
I’ll set up more tests to verify

Jatinbhardwaj-093 committed:
fix the truncation. Added better test case

fix test case

Jatinbhardwaj-093 commented on sympy/polys/tests/test_densetools.py:
```diff
@@ -282,6 +286,30 @@ def test_dmp_revert():
+        if f[-2] == 0:
+            f[-2] = QQ(1, 1)
+
+        g = dup_reversion(f, deg, QQ)
+        composed = dup_slice(dup_compose(f, g, QQ), 0, deg, QQ)
+
+        expected = [QQ(1), QQ(0)]
+        assert composed == expected
```
Since we only need x after slicing, specifying the expected type as x should sufficient, no truncation is necessary.

oscarbenjamin commented:
Here is a test:
```python
from sympy import *
from random import randint
from sympy.polys.densetools import dup_reversion, dup_compose

Qrand = lambda: QQ(randint(-10, 10), randint(1, 10))

for n1 in range(0, 5):
    for m1 in range(0, 10):
        for _ in range(100):
            p1 = [Qrand() for i in range(n1)]
            if len(p1) < 2 or p1[-2] == 0:
                continue
            if p1:
                p1[-1] = QQ(0)
            p2 = dup_compose(dup_reversion(p1, n1, QQ), p1, QQ)
            got = p2[::-1][:n1]
            while got[-1] == 0:
                got.pop()
            assert got == [0, 1]
            print('fine:', p1)
```
Are there any other `dup_*` functions that do series methods like this? I wonder if there should be a naming convention for them like `dup_series_reversion` rather than `dup_reversion`.

Jatinbhardwaj-093 commented:
My implementation is somewhat similar. If needed, I can just increase the number of trials.

> Are there any other dup_* functions that do series methods like this? I wonder if there should be a naming convention for them like dup_series_reversion rather than dup_reversion.

We could add functions like mul and pow as well, but I believe those are already specific to series, so it’s fine to keep them in the series file.
As for the naming convention, I’m not sure it’s necessary for reversion in particular — I think dup_reversion is fine, since it can also be applied to polynomials.

That said, if we really want to, we could define specific functions that output terms modulo x**n as series. Then we could group common operations — such as mul, pow, reversion, inversion, compose, and others — under a consistent prefix like dup_series_*.

oscarbenjamin commented:
I'm not sure what makes sense for organising the code with `dup_*` or `dup_series_*` functions here but we should aim for the code to be organised consistently.

Jatinbhardwaj-093 committed:
change name dup_reversion -> dup_series_reversion. Added _dup_random helper func to generate random dup element

oscarbenjamin commented:
I see also there are fast algorithms for this:
https://fredrikj.net/math/reversion.pdf

oscarbenjamin commented on sympy/polys/tests/test_densetools.py:
```diff
@@ -282,6 +292,32 @@ def test_dmp_revert():
+        if f[-2] == 0:
+            f[-2] = QQ(1, 1)
+
+        g = dup_series_reversion(f, deg, QQ)
+        composed = dup_slice(dup_compose(f, g, QQ), 0, deg, QQ)
```
Maybe it makes sense to have a `dup_series_compose` function.

oscarbenjamin commented on sympy/polys/densetools.py:
```diff
@@ -1431,3 +1432,68 @@ def dmp_revert(f, g, u, K):
+    Computes the compositional inverse of f using Newton iteration.
+    The result is computed modulo x**n.
+
+    Algorithm
+    =========
+
+    Given $f(x) = a_1 x + a_2 x^2 + \dots$ with $a_1$ invertible.
+
+    The goal is to find $g(x)$ such that:
+    $
+    f(g(x)) = x \mod x^n
+    $
+
+    The algorithm uses iterative refinement where $g_k$ is the approximation modulo $x^k$.
+
+    Initial approximation:
+    $
+    g_1(x) = x/a_1
+    $
+
+    Iterative step for $k = 2, 3, \dots, n$:
+    $
+    \delta = f(g_{k-1}(x)) - x \mod x^k
+    $
+    $
+    g_k(x) = g_{k-1}(x) - \delta/a_1 \mod x^k
+    $
+
+    This process incrementally builds the correct coefficients of the compositional inverse.
```
I'm not sure that the algorithm description is needed in the docstring.

It could be noted that it is not necessarily a fast algorithm for large `n`.

Jatinbhardwaj-093 committed:
added more dup_series_* funcs to perform series operation

Jatinbhardwaj-093 commented:
No review needed at the moment. I still need to clean up dup_series_compose.
Soon, I’ll also look into implementing or analyzing the fast algorithm for series reversion.

Jatinbhardwaj-093 commented:
I’m trying to implement the fast Lagrange inversion algorithm, and here’s what I have so far:
```python
def dup_series_reversion2(f, n, K):
    if not f:
        return []

    if len(f) < 2 or f[-1] != K.zero:
        raise ValueError("f must have zero constant term")
    if f[-2] == K.zero:
        raise ValueError("f must have nonzero linear term")

    # Step 0: h = x / f mod x**(n-1)
    f1 = dup_rshift(f, 1, K)
    f1_inv = dup_revert(f1, n, K)
    h = dup_lshift(f1_inv, 2, K)
    h = dup_truncate(h, n, K)

    m = _ceil(sqrt(n - 1))

    # Precompute powers of h: h, h^2, ..., h^m mod x**(n-1)
    H = [h]
    for _ in range(1, m):
        h_prev = H[-1]
        h_next = dup_series_mul(h_prev, h, n, K)
        H.append(h_next)

    # Initialize g with n zeros
    g = [K.zero] * n

    # First block: compute g[i] for i = 1 to m - 1
    for i in range(1, m):
        coeff = dup_nth(H[i], i - 1, K)  # coeff = [x**(i-1)](h^i)
        # (1/i) * [x**(i-1)](h^i)
        g[-(i + 1)] = K.quo(coeff, K(i))

    # t = h^m
    t = H[m - 1]

    # Loop over blocks of size m
    for i in range(m, n, m):
        # g[i] = (1 / i) * [x**(i-1)](t)
        coeff = dup_nth(t, i - 1, K)
        g[-(i + 1)] = K.quo(coeff, K(i))

        # Now fill g[i + j] for 1 <= j < m
        for j in range(1, m):
            if i + j >= n:
                break

            s = K.zero
            for k in range(0, i + j):
                c1 = dup_nth(t, k, K)
                c2 = dup_nth(H[j], i + j - k - 1, K)
                s += c1 * c2

            g[-(i + j + 1)] = K.quo(s, K(i + j))

        # t = t * h^m mod x**(n-1)
        t = dup_series_mul(t, H[m - 1], n, K)

    return dup_strip(g)
```

However, this doesn’t seem to produce the correct reversion. I’m not sure where the issue lies after this point. If you can help me identify what might be going wrong, that would be great.

Jatinbhardwaj-093 committed:
fixes dup_series_mul, dup_series_compose

use strip at end

revert reversion2 funciton

Jatinbhardwaj-093 committed:
make use of series_compose in dup_series_reversion

Jatinbhardwaj-093 commented:
This is ready for review.

`dup_series_mul` currently uses the Cauchy product with immediate truncation. Should I consider adding Karatsuba multiplication as an optimization for larger inputs, applying truncation after multiplication?

oscarbenjamin commented:
> Should I consider adding Karatsuba multiplication as an optimization for larger inputs, applying truncation after multiplication?

Yes. You could do some timings but probably the cutoff is not much different as for `dup_mul`.

oscarbenjamin commented:
> Should I consider adding Karatsuba multiplication as an optimization for larger inputs, applying truncation after multiplication?

Worth noting that I think the "fast" reversion algorithm is only faster if you have subquadratic polynomial multiplication i.e. only if using larger polynomials and using something like Karatsuba.

oscarbenjamin commented:
The idea of testing with randomly generated inputs works for all of the methods e.g. you can test `dup_series_pow` by comparing with `dup_pow` and so on.

I think it is good to have a few explicit tests for some base cases and typical examples but then random tests give better coverage.

oscarbenjamin commented on sympy/polys/tests/test_densetools.py:
```diff
@@ -282,6 +292,32 @@ def test_dmp_revert():
 
 
+def test_dup_series_reversion(trials=50):
+    for _ in range(trials):
+        deg = randint(2, 10)
+        f = _dup_random(-10, 10, deg, QQ)
+        if len(f) < 2:
+            continue
+
+        f[-1] = QQ(0)
+        if f[-2] == 0:
+            f[-2] = QQ(1, 1)
+
+        g = dup_series_reversion(f, deg, QQ)
```
In all of the cases the same `deg` is used for `_dup_random` and `dup_series_reversion`. I think it is good to have cases where they are not equal as well.

Jatinbhardwaj-093 commented:
You’re right — it should be:

> I guess it should be:
    # Step 0: h = x / f mod x**(n-1)
    f1 = dup_rshift(f, 1, K)
    h = dup_revert(f1, n-1, K)


Also, I was mistakenly fetching the wrong nth coefficient both should be one lesser.
Just tested — it’s working correctly now.

Importantly verified too.

oscarbenjamin commented on sympy/polys/densearith.py:
```diff
@@ -997,6 +1031,46 @@ def dmp_pow(f, n, u, K):
+
+def dup_series_pow(f, n, prec, K):
+    """
+    Raise ``f`` to the ``n``-th power in ``K[[x]]`` up to ``x**prec``.
```
It is not completely clear what `x**prec` means e.g. does it mean:
```
a + b*x + ... + c*x**prec + O(x**(prec + 1))
```
I think it is better to say `mod x**prec` or something,

oscarbenjamin commented on sympy/polys/densetools.py:
```diff
@@ -1026,6 +1028,54 @@ def dmp_compose(f, g, u, K):
+    """
+    Compute f(g) mod x^n using divide and conquer composition.
+
+    This function efficiently computes the first n terms of the composition f(g(x)).
```
The "first n terms" is ambiguous because what if some coefficients are zero?

Jatinbhardwaj-093 committed:
added fast lagrange inversion for 'reversion'. added karastuba for series_mul. Improve test cases

code quality fix

Jatinbhardwaj-093 committed:
improve test cases

oscarbenjamin commented on sympy/polys/densebasic.py:
```diff
@@ -1841,6 +1841,23 @@ def dmp_slice(f, m, n, u, K):
+    >>> from sympy.polys.densebasic import dup_truncate
+
+    >>> dup_truncate([1, 2, 3, 4, 5], 3, ZZ)
+    [3, 4, 5]
```
The inputs should be ZZ(1), ZZ(2) etc. You can use `ZZ.map` though.

These doctests would be more understandable if they displayed outputs that look like polynomials. There is no `dup_print` function but if there was it could be:
```python
>>> p = ZZ.map([1, 2, 3, 4, 5])
>>> p3 = dup_truncate(p, 3, ZZ)
>>> dup_print(p, ZZ)
x**4 + 2*x**3 + 3*x**2 + 4*x + 5
>>> dup_print(p3, ZZ)
3*x**2 + 4*x + 5
```
Perhaps even `ZZ.map` could be replaced with `p = dup_from_list([1, 2, 3, 4, 5], ZZ)`

oscarbenjamin commented on sympy/polys/densetools.py:
```diff
@@ -1431,3 +1479,82 @@ def dmp_revert(f, g, u, K):
         raise MultivariatePolynomialError(f, g)
+
+
+def dup_series_reversion(f, n, K):
+    r"""
```
Is this version of the reversion algorithm always more efficient than the previous one? Have you done any timings?

I would not be surprised if this implementation is slower for small `n`. It can only be asymptotically faster in the regime where Karatsuba kicks in which means you need a good size `n`.

I think it would be good to keep both versions and to have something that chooses which one to use e.g.:
```
def dup_series_reversion(...):
    if len(f) > 100:
        return _dup_series_lagrange
   else:
        return ...
```

oscarbenjamin replied:
From some quick timings it seems that maybe the Lagrange version is only slower for `n < 4` so the other version is really only useful for very small inputs. It could perhaps be better optimised for that case then. 

For very small inputs we can hard-code the answer so perhaps it is also good to have something for the really small cases:
```python
In [45]: from sympy.polys.densetools import dup_series_reversion

In [46]: a, b, c = symbols('a, b, c')

In [47]: F = ZZ.frac_field(a,b,c)

In [48]: p = F.map([a, b, c, 0])

In [49]: dup_series_reversion(p, 4, F)
Out[49]: [(-a*c + 2*b**2)/(c**5), -b/(c**3), 1/c, 0]
```
For `n = 4` it is:
```python
def dup_series_reversion_4(f, K):
    a, b, c, d = f
    if not K.is_zero(d):
        raise ValueError("f must have zero constant term")
    cinv = K.revert(c)
    cinv2 = cinv**2
    cinv3 = cinv*cinv2
    cinv5 = cinv3*cinv2
    return [(2*b**2 - a*c)*cinv5, -b*cinv3, cinv, K.zero]
```
That is 60x faster than the lagrange version at this size:
```python
In [60]: %timeit dup_series_reversion_4(p, QQ)
3.22 μs ± 38.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [62]: %timeit dup_series_reversion(p, 4, QQ)
171 μs ± 3.79 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```
The smaller cases are:
```python
In [64]: p = F.map([b, c, 0])

In [65]: dup_series_reversion(p, 3, F)
Out[65]: [-b/(c**3), 1/c, 0]

In [66]: dup_series_reversion(p, 2, F)
Out[66]: [1/c, 0]

In [67]: dup_series_reversion(p, 1, F)
Out[67]: []
```
That last case is returning 0 which should be an error.

Jatinbhardwaj-093 commented:
> If you have e.g. dup_series_revert_lagrange and dup_series_revert_simple or something then it is easier to test the implementations by comparing them to each other with randomly generated inputs.

I’m not entirely sure about this — if we try to implement the original Lagrange inversion using full multiplication, then relying on dup_mul will naturally slow things down quite a bit. Alternatively, if we try to create a “simple” reversion using previous Newton iteration that going to use series_compose, which also going to end up using dup_series_mul internally.

oscarbenjamin commented:
I think that you want to be able to compare `dup_series_mul` with something else e.g.:
```
def _dup_series_mul_basecase(...): ...

def _dup_series_mul_karatsuba(...): ...

def dup_series_mul(...): ...
```
This pattern is used a lot in FLINT:

- make a simple and obvious correct implementation
- make a more optimised implementation
- make tests that compare the output of the two implementations

It is much easier to test things like larger inputs if you can write tests like this based on randomly generated inputs.

Jatinbhardwaj-093 committed:
fixes karatsuba algo. Seperate base and karatsuba algo for mul.
Added base case upto n<=4 for fast reversion compute.
Improve test coverage for all to operation to reach and use karatsuba mul.

Jatinbhardwaj-093 committed:
Added dup_from_list and dup_print.
Improve doctest.

oscarbenjamin commented on sympy/polys/densebasic.py:
```diff
@@ -1885,3 +1904,75 @@ def dup_random(n, a, b, K):
+    >>> from sympy.polys.domains import QQ
+    >>> from sympy.polys.densebasic import dup_from_list
+
+    >>> dup_from_list([1, 0, 5, 0, 7], QQ)
+     [1, 0, 5, 0, 7]
```
This is indented slightly.

It would be good to clarify here something like:
```
>>> QQ.of_type(1)
False
>>> QQ.of_type(p[0])
True
```

Jatinbhardwaj-093 committed:
minor improvements

Jatinbhardwaj-093 committed:
added hypothesis test

oscarbenjamin commented on sympy/polys/densearith.py:
```diff
@@ -832,6 +833,91 @@ def dmp_mul(f, g, u, K):
+    if not f or not g or n <= 0:
+        return []
+
+    f = dup_truncate(f, n, K)[::-1]
+    g = dup_truncate(g, n, K)[::-1]
+
+    h = [K.zero] * n
+    for i in range(len(f)):
+        for j in range(min(len(g), n - i)):
+            h[i + j] += f[i] * g[j]
```
If this was `min(len(f), n)` then we would not need `dup_truncate`.

oscarbenjamin commented on sympy/polys/tests/test_densearith.py:
```diff
@@ -679,6 +687,29 @@ def test_dmp_mul():
+        f = _dup_random(-20, 0, deg, ZZ)
+        g = _dup_random(0, 20, randint(1, 200), ZZ)
+
+        mul = dup_series_mul(f, g, deg, ZZ)
+        expected = dup_truncate(dup_mul(f, g, ZZ), deg, ZZ)
+        assert mul == expected
```
Maybe let's replace the random tests here with deterministic tests that give full line coverage like
```
f = range(200)
g = [1] * 200

f = [n**2 for n in range(200)]
```
We want one or two small tests written explicitly but for large tests we can just compare dup_series_mul and dup_mul.

Jatinbhardwaj-093 committed:
100% coverage

Jatinbhardwaj-093 committed:
hypothesis testing for dup rationals, Proper division b/w different dup needed

added few more tests

fix issue

oscarbenjamin commented:
Let me know when this is ready to go.

Does it have 100% coverage from deterministic tests?

oscarbenjamin commented:
Okay, looks good.
